coding/coding.go

```diff
@@ -1,75 +1,112 @@
-package ipfsld
+package coding
 
 import (
-	mc "github.com/jbenet/go-multicodec"
-	mccbor "github.com/jbenet/go-multicodec/cbor"
-	mcmux "github.com/jbenet/go-multicodec/mux"
+	"bytes"
+	"fmt"
+	"io"
 
-	ipld "github.com/ipfs/go-ipld"
+	cbor "github.com/ipfs/go-ipld/coding/cbor"
+	json "github.com/ipfs/go-ipld/coding/json"
 	pb "github.com/ipfs/go-ipld/coding/pb"
+
+	memory "github.com/ipfs/go-ipld/memory"
+	stream "github.com/ipfs/go-ipld/stream"
+	mc "github.com/jbenet/go-multicodec"
+)
+
+var Header []byte
+
+const (
+	HeaderPath = "/mdagv1"
 )
 
-// defaultCodec is the default applied if user does not specify a codec.
-// Most new objects will never specify a codec. We track the codecs with
-// the object so that multiple people using the same object will continue
-// to marshal using the same codec. the only reason this is important is
-// that the hashes must be the same.
-var defaultCodec string
+type Codec int
 
-var muxCodec *mcmux.Multicodec
+const (
+	NoCodec       Codec = 0
+	CodecProtobuf Codec = iota
+	CodecCBOR
+	CodecJSON
+	CodecCBORNoTags
+)
+
+var StreamCodecs map[string]func(io.ReadSeeker) (stream.NodeReader, error)
```

Comment:
Go is not a functional language. It makes it harder to reason about the code to have `funcs` all over the place. I think these should be an interface, and one that follows the `io` pattern:

```go
type StreamCodec interface {
  // read first object from stream and populate o
  Read(stream io.ReadSeeker, stream.NodeReader) error
 
  // write o into stream 
  Write(stream io.ReadSeeker, stream.NodeWriter) error
}
```
and really, why `NodeReader`? it should probably be as simple as just a `ipld.Node` type.